File: kafkaStreamsModelServer/src/main/java/com/lightbend/java/kafkastreams/modelserver/memorystore/ModelProcessor.java
package com.lightbend.java.kafkastreams.modelserver.memorystore;

import com.lightbend.java.kafkastreams.store.StoreState;
import com.lightbend.java.model.CurrentModelDescriptor;
import com.lightbend.java.model.DataConverter;
import com.lightbend.java.model.ModelServingInfo;
import com.lightbend.java.model.ModelWithDescriptor;
import org.apache.kafka.streams.processor.AbstractProcessor;
import org.apache.kafka.streams.processor.ProcessorContext;

import java.util.Optional;

/**
 * Created by boris on 7/12/17.
 */

public class ModelProcessor extends AbstractProcessor<byte[], byte[]> {

    private StoreState modelStore;

    @Override
    public void process(byte[] key, byte[] value) {

        Optional<CurrentModelDescriptor> descriptor = DataConverter.convertModel(value);
        if(!descriptor.isPresent()){
            return;                                                 // Bad record
        }
        Optional<ModelWithDescriptor> modelWithDescriptor = DataConverter.convertModel(descriptor);
        if(modelWithDescriptor.isPresent()){
           modelStore.setNewModel(modelWithDescriptor.get().getModel());
           modelStore.setNewServingInfo(new ModelServingInfo(modelWithDescriptor.get().getDescriptor().getName(),
                    modelWithDescriptor.get().getDescriptor().getDescription(), 0));
        }
    }

    @Override
    public void init(ProcessorContext context) {
        modelStore = StoreState.getInstance();
    }
}

File: kafkaStreamsModelServer/src/main/scala/com/lightbend/scala/kafkastreams/modelserver/memorystore/ModelProcessor.scala
package com.lightbend.scala.kafkastreams.modelserver.memorystore

import com.lightbend.scala.kafkastreams.store.StoreState
import com.lightbend.scala.modelServer.model.{ModelToServe, ModelToServeStats, ModelWithDescriptor}
import org.apache.kafka.streams.processor.{AbstractProcessor, ProcessorContext}

import scala.util.Success

/**
 * Handle new model parameters; updates the current model used for scoring.
 */
class ModelProcessor extends AbstractProcessor[Array[Byte], Array[Byte]]{

  private var modelStore: StoreState = null

  override def process (key: Array[Byte], value: Array[Byte] ): Unit = {

    ModelToServe.fromByteArray(value) match {
      case Success(descriptor) => {
        ModelWithDescriptor.fromModelToServe(descriptor) match {
          case Success(modelWithDescriptor) => {
            modelStore.newModel = Some(modelWithDescriptor.model)
            modelStore.newState = Some(ModelToServeStats(descriptor))
          }
          case _ => // error; ignore
        }
      }
      case _ => // error; ignore
    }
  }

  // Exercise:
  // Like all good production code, we're ignoring errors ;) in two places. That is, we filter to keep
  // messages where a `Success(x)` is returned true and ignore the `Failure(exception)` results.
  // With the topology API, this is harder to fix; what could you do in those `case _ =>` clauses??
  // See the implementation of `DataRecord`, where we inject fake errors. Add the same logic to `ModelToServe` and
  // `ModelWithDescriptor`.

  override def init(context: ProcessorContext): Unit = {
    modelStore = StoreState()
  }
}
